### DISCUSSION TITLE
Option to disable auth key expiration

### DISCUSSION THREAD

**Discussion by rakema01** [torrust/torrust-tracker#244] [Feature requests]
Would be a neat thing to have when the focus of the site is to seed for long periods of time. Also especially helpful when you've got too many torrents and having to change the tracker URL to match the key is a pain. Of course, there's the option to change the expiry length to a random huge number but having an option to outright disable expiry checks is a bit cleaner in my opinion. In its place, we could have an option for the key be manually regenerated when needed.

**Comment by mickvandijke** [comment]
Good idea. I will make it so that if valid_until is set to 0, the key won't expire.

**Reply by josecelano** [reply]
Hey @WarmBeer after the latest refactorings we have:

- [Keys](https://github.com/torrust/torrust-tracker/blob/develop/src/tracker/auth.rs#L86)
- [ExpiringKeys](https://github.com/torrust/torrust-tracker/blob/develop/src/tracker/auth.rs#L55-L58)

```rust
pub struct Key(String);

pub struct ExpiringKey {
    pub key: Key,
    pub valid_until: DurationSinceUnixEpoch,
}
```

Maybe what we could do if to add a new collection of permanent keys to the tracker:

Current tracker attributes:

```rust
pub struct Tracker {
    pub config: Arc<Configuration>,
    mode: TrackerMode,
    keys: RwLock<std::collections::HashMap<Key, auth::ExpiringKey>>,
    whitelist: RwLock<std::collections::HashSet<InfoHash>>,
    torrents: RwLock<std::collections::BTreeMap<InfoHash, torrent::Entry>>,
    stats_event_sender: Option<Box<dyn statistics::EventSender>>,
    stats_repository: statistics::Repo,
    pub database: Box<dyn Database>,
}
```

New struct after the change:

```rust
pub struct Tracker {
    // ...
    expiring_keys: RwLock<std::collections::HashMap<Key, auth::ExpiringKey>>,
    permanent_keys: RwLock<std::collections::HashMap<Key, auth::ExpiringKey>>,
    // ...
}
```

We also need to change the API auth endpoint to reflect that change. We can use a single resource again where the expiry date is optional or we could have two different resources:

- `/api/v1/keys/permanent`
- `/api/v1/keys/expiring`

I've been working on that endpoint today: [PR](https://github.com/torrust/torrust-tracker/pull/243).


Alternatively, we could change the name again and allow the `expiry_time` to be optional:

```rust
pub struct Key(String);

pub struct RegisteredKey {
    pub key: Key,
    pub maybe_valid_until: Option<DurationSinceUnixEpoch>,
}
```

cc @da2ce7

**Comment by josecelano** [comment]
This was implemented [here](https://github.com/torrust/torrust-tracker/issues/978).
